### PR TITLE
Create specific price when product price is updated in order

### LINKED ISSUE #19890: Update product's price in the order does not update the discount
Be aware of the EPIC - BO - Display and update error on cart rule in the order #19594 to solve this issue

#### Describe the bug

On the back-office, when I create a percentage discount and then change the product's price on the order, the discount is not updated.

#### Expected behavior

The discount should apply the 50% reduction on all the product amount even if after they are updated 

#### Steps to Reproduce

1. Add the product to any order
2. Create a cart rule of 50% in the order page
3. edit the product price
4. See error - wrong discount total

**Screenshots**

#### Additional information

* PrestaShop version: N/A
* PHP version: N/A

### PR DESCRIPTION
**Pull request by sowbiba**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | Fix discount calculation when product price in order is updated. We create a specific price like in AddProductToOrder to fix products total
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #19890
| How to test?  | Create a cart rule of 50% on selected product<br>Add the discounted product to any order and see the discount <br>edit the product price<br>The discount must be correct

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19970)

### CODE REVIEW

**Review by jolelievre — DISMISSED** [review]
I think we would need to `restoreOrderProductsSpecificPrices` for update and delete actions as well if we want the prices to be correctly set when the cart rules a recomputed
But since we plan to change this and stop removing them in a coming PR I guess we can wait for the next one

**Comment by Progi1984** [comment]
@sowbiba Travis is not happy :red_circle: 

```
Fatal error: A function with return type must return a value (did you mean "return null;" instead of "return;"?) in /home/travis/build/PrestaShop/PrestaShop/src/Adapter/Order/AbstractOrderHandler.php on line 148
```

**Comment by matthieu-rolland** [comment]
> @sowbiba Travis is not happy 🔴
> 
> ```
> Fatal error: A function with return type must return a value (did you mean "return null;" instead of "return;"?) in /home/travis/build/PrestaShop/PrestaShop/src/Adapter/Order/AbstractOrderHandler.php on line 148
> ```

It's weird, he should be able to return null in this function since it has a nullable return type.

**Review by matthieu-rolland — COMMENTED** [review]
LGTM, I'll aprove this PR once the ci issue is fixed

File: src/Adapter/Order/AbstractOrderHandler.php [hunk 1]
```diff
@@ -103,4 +112,167 @@ protected function getPrecisionFromCart(Cart $cart): int
...
+        return $combination;
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param int $langId
+     *
+     * @return Product
+     */
+    protected function getProductObject(ProductId $productId, $langId)
```
Comment (Progi1984) [inline src/Adapter/Order/AbstractOrderHandler.php]:
```suggestion
    protected function getProduct(ProductId $productId, $langId)
```

File: src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php [hunk 2]
```diff
@@ -63,60 +66,90 @@ public function handle(UpdateProductInOrderCommand $command)
...
+                $orderDetail->unit_price_tax_incl = $product_price_tax_incl;
+
+                $orderDetail->total_price_tax_incl += $diff_price_tax_incl;
+                $orderDetail->total_price_tax_excl += $diff_price_tax_excl;
+
+                $cart = Cart::getCartByOrderId($order->id);
+                if (!($cart instanceof Cart)) {
+                    throw new OrderException('Cart linked to the order cannot be found.');
+                }
+                $product = $this->getProductObject(new ProductId((int) $orderDetail->product_id), (int) $order->id_lang);
```
Comment (Progi1984) [inline src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php]:
```suggestion
                $product = $this->getProduct(new ProductId((int) $orderDetail->product_id), (int) $order->id_lang);
```

**Review by Progi1984 — COMMENTED** [review]
A small rename

File: src/Adapter/Order/AbstractOrderHandler.php (hunk 1 above)
Comment (sowbiba) [reply src/Adapter/Order/AbstractOrderHandler.php]: done. I renamed getOrderObject too. That introduce modifications in multiple handlers ;)

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by matthieu-rolland — APPROVED** [review]

**Review by jolelievre — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hi @sowbiba !

Your PR is QA :heavy_check_mark: .

The voucher is updated when product price is updated, see screenshots : 

Before updating product price : 
pr19970 before update

After updating product price : 
pr19970 after update


Thanks !

**Comment by jolelievre** [comment]
Thank you @sowbiba @Robin-Fischer-PS and @SD1982

**Pull request merged**